File: qabelbox/src/main/java/de/qabel/qabelbox/storage/notifications/AndroidStorageNotificationManager.java
package de.qabel.qabelbox.storage.notifications;

import javax.inject.Inject;

import de.qabel.box.storage.BoxFile;
import de.qabel.qabelbox.storage.model.BoxUploadingFile;

public class AndroidStorageNotificationManager implements StorageNotificationManager {

    private StorageNotificationPresenter presenter;
    private BoxUploadingFile lastUpload;

    @Inject
    public AndroidStorageNotificationManager(StorageNotificationPresenter storageNotificationPresenter) {
        this.presenter = storageNotificationPresenter;
    }

    @Override
    public void updateUploadNotification(int queueSize, BoxUploadingFile currentUpload) {
        BoxUploadingFile displayFile;
        if (currentUpload != null) {
            this.lastUpload = currentUpload;
            displayFile = currentUpload;
        } else {
            displayFile = lastUpload;
        }
        presenter.updateUploadNotification(queueSize,
                new StorageNotificationInfo(displayFile.getName(), displayFile.getPath(),
                        displayFile.getOwnerIdentifier(), displayFile.uploadedSize,
                        displayFile.totalSize));
    }

    @Override
    public BoxTransferListener addDownloadNotification(final String ownerKey, final String path, final BoxFile file) {
        return new BoxTransferListener() {

            private StorageNotificationInfo notificationInfo =
                    new StorageNotificationInfo(file.getName(), path, ownerKey, 0, file.getSize());

            @Override
            public void onProgressChanged(long bytesCurrent, long bytesTotal) {
                notificationInfo.setProgress(bytesCurrent, bytesTotal);
                presenter.updateDownloadNotification(notificationInfo);
            }

            @Override
            public void onFinished() {
                notificationInfo.complete();
                presenter.updateDownloadNotification(notificationInfo);
            }
        };
    }
}

File: qabelbox/src/main/java/de/qabel/qabelbox/box/notifications/AndroidStorageNotificationManager.kt
package de.qabel.qabelbox.box.notifications

import de.qabel.client.box.interactor.FileOperationState
import de.qabel.client.box.interactor.FileOperationState.Status
import javax.inject.Inject

class AndroidStorageNotificationManager
@Inject constructor(private val presenter: StorageNotificationPresenter) : StorageNotificationManager {

    override fun updateUploadNotification(fileOperationState: FileOperationState) {
        val info = fileOperationState.toNotificationInfo()
        when (fileOperationState.status) {
            Status.PREPARE -> presenter.showEncryptingUploadNotification(info)
            Status.LOADING,
            Status.COMPLETING -> presenter.showUploadProgressNotification(info)
            Status.COMPLETE -> presenter.showUploadCompletedNotification(info)
            Status.ERROR -> presenter.showUploadFailedNotification(info)
            Status.HIDDEN,
            Status.CANCELED -> presenter.cancelNotification(info)
        }
    }

    override fun updateDownloadNotification(fileOperationState: FileOperationState) {
        val info = fileOperationState.toNotificationInfo()
        when (fileOperationState.status) {
            Status.PREPARE,
            Status.LOADING -> presenter.showDownloadProgressNotification(info)
            Status.COMPLETING -> presenter.showDecryptingDownloadNotification(info)
            Status.COMPLETE -> presenter.showDownloadCompletedNotification(info)
            Status.ERROR -> presenter.showDownloadFailedNotification(info)
            Status.HIDDEN,
            Status.CANCELED -> presenter.cancelNotification(info)
        }
    }

    private fun FileOperationState.toNotificationInfo(): StorageNotificationInfo =
            StorageNotificationInfo(entryName, path.toString(), ownerKey.publicKey,
                    time, done, size)

}
